contracts/RewardPool.sol

```diff
@@ -256,6 +289,179 @@ contract RewardPool is AccessControl, ReentrancyGuard {
     emit Withdraw(amount);
   }
 
+  /**
+   * @dev Internal function to get the period id
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @return bytes32 period id
+   */
+  function _getPeriodId(
+    uint48 periodStart,
+    uint48 periodEndExclusive
+  ) internal pure returns (bytes32) {
+    if (periodStart >= periodEndExclusive)
+      revert PeriodInvalid(periodStart, periodEndExclusive);
+
+    return keccak256(abi.encode(periodStart, periodEndExclusive));
+  }
+
+  /**
+   * @dev Allows the owner to update kpis for a reward period
+   * @param kpis Array of kpis to add
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @param kpiFunctionId Bytes32 identifier of the kpi function (e.g., github commit hash)
+   * @notice Allowed only for address with DEFAULT_ADMIN_ROLE
+   */
+  function updatePeriodKpis(
+    IRewardFunction.Kpi[] calldata kpis,
+    uint48 periodStart,
+    uint48 periodEndExclusive,
+    bytes32 kpiFunctionId
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    bytes32 periodId = _getPeriodId(periodStart, periodEndExclusive);
+
+    PeriodData storage period = _periods[periodId];
+
+    if (period.processedAt != 0) revert PeriodAlreadyProcessed(periodId);
+
+    for (uint256 i = 0; i < kpis.length; i++) {
+      if (kpis[i].referrerAddress == address(0))
+        revert ZeroAddressNotAllowed(i);
+
+      period.users.add(kpis[i].referrerAddress);
+      period.kpis[kpis[i].referrerAddress] = kpis[i].kpi;
+
+      emit KpiUpdated(
+        kpis[i].referrerAddress,
+        periodId,
+        kpis[i].kpi,
+        periodStart,
+        periodEndExclusive,
+        kpiFunctionId
+      );
+    }
+  }
+
+  /**
+   * @dev Allows the owner to process a reward period and distribute rewards
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @param totalRewardAmount Total amount of rewards to distribute
+   * @notice Allowed only for address with DEFAULT_ADMIN_ROLE
+   */
+  function processPeriod(
+    uint48 periodStart,
```

Comment:
Probably not a big deal, but there's no way to inspect, just by looking at contract state, what periods have had data uploaded. We can query about particular start/end ranges, but we can't retrieve e.g., a list of periods that are available. With indexing we can look at `KpiUpdated` events, but there's nothing public on the contract that can tell us what's been uploaded.

Not sure if this is a big deal (probably not), but wanted to call it out. It's probably not worth storing the extra state required to return this data.

Reply:
Yeah this is something that we can add in later, seems like it would tie in well with pre-registering the periods before campaign start.